controllers/restart_incompatible_processes.go

```diff
@@ -0,0 +1,116 @@
+ */
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	fdbv1beta2 "github.com/FoundationDB/fdb-kubernetes-operator/api/v1beta2"
+	"github.com/FoundationDB/fdb-kubernetes-operator/internal"
+)
+
+// restartIncompatibleProcesses is a reconciler that will restart incompatible fdbserver processes, this can happen
+// during an upgrade when the kill command doesn't reach all processes, see: https://github.com/FoundationDB/fdb-kubernetes-operator/issues/1281
+type restartIncompatibleProcesses struct{}
+
+// reconcile runs the reconciler's work.
+func (restartIncompatibleProcesses) reconcile(ctx context.Context, r *FoundationDBClusterReconciler, cluster *fdbv1beta2.FoundationDBCluster) *requeue {
+	logger := log.WithValues("namespace", cluster.Namespace, "cluster", cluster.Name, "reconciler", "restartIncompatibleProcesses")
+
+	pods, err := r.PodLifecycleManager.GetPods(ctx, r, cluster, internal.GetPodListOptions(cluster, "", "")...)
+	if err != nil {
+		return &requeue{curError: err}
+	}
+
+	podMap := internal.CreatePodMap(cluster, pods)
+
+	adminClient, err := r.getDatabaseClientProvider().GetAdminClient(cluster, r.Client)
+	if err != nil {
+		return &requeue{curError: err}
+	}
+	defer adminClient.Close()
+
+	status, err := adminClient.GetStatus()
+	if err != nil {
+		return &requeue{curError: err}
+	}
+
+	if len(status.Cluster.IncompatibleConnections) == 0 {
+		return nil
+	}
+
+	logger.Info("incompatible connections", "incompatibleConnections", status.Cluster.IncompatibleConnections)
+	incompatibleConnections := map[string]fdbv1beta2.None{}
+	for _, incompatibleAddress := range status.Cluster.IncompatibleConnections {
+		address, err := fdbv1beta2.ParseProcessAddress(incompatibleAddress)
+		if err != nil {
+			logger.Error(err, "could not parse address in incompatible connections", "address", incompatibleAddress)
+			continue
+		}
+
+		incompatibleConnections[address.IPAddress.String()] = fdbv1beta2.None{}
+	}
+
+	var hasRestart bool
+	for _, processGroup := range cluster.Status.ProcessGroups {
+		pod, ok := podMap[processGroup.ProcessGroupID]
+		if !ok || pod == nil {
+			logger.V(1).Info("Could not find Pod for process group ID",
+				"processGroupID", processGroup.ProcessGroupID)
+			continue
+		}
+
+		if isIncompatible(incompatibleConnections, processGroup) {
+			logger.Info("restart process group with incompatible version", "processGroupID", processGroup.ProcessGroupID)
+			curErr := r.restartFdbserverProcess(pod.Name, pod.Namespace)
+			if curErr != nil {
+				err = fmt.Errorf("could not restart %s/%s with error: %w", pod.Name, pod.Namespace, curErr)
+			}
+			continue
+		}
+
+	}
+
+	if err != nil {
+		return &requeue{curError: err, delay: 15 * time.Second}
+	}
+
+	// When we restarted the incompatible fdbserver processes we have to wait until they report to the cluster again to
+	// prevent to restart them in a loop.
+	if hasRestart {
+		return &requeue{message: "wait until processes are restarted", delay: 15 * time.Second}
```

Comment:
I think we actually don't need to requeue here otherwise we might be in an endless loop.